Add multi-threading compression

Cyan4973 opened the pull request:
This patch adds multi-threading support for compression to the `lz4` CLI.

Features :
- Display if the binary supports multithreading or not (welcome message) 
- Automatically select a nb of threads, based on local cpu
  + can also be controlled manually
- Provides a detailed summary in verbose mode, featuring speed and cpu utilization

Here are a few results using this patch : 

| cpu | corpus | c.level | v1.9.4 | this PR | speedup |
| --- | --- | --- | --- | --- | --- |
| i7-9700 | enwik9 | 1 | 2.3 s | 0.45 s | x5.1 |
| i7-9700 | silesia.tar | 9 | 5.5 s | 1.0 s | x5.5 |
| i7-9700 | silesia.tar | 12 | 17 s | 3.0 s | x5.6 |
| i7-9700 | enwik9 | 9 | 23 s | 4.1 s | x5.6 |
| M1 Pro | enwik9 | 1 | 2.2 s | 0.42 s | x5.2 |
| M1 Pro | silesia.tar | 9 | 5.1 s | 0.75 s | x6.8 |
| M1 Pro | silesia.tar | 12 | 16.5 s | 2.5 s | x6.6 |
| M1 Pro | enwik9 | 9 | 21 s | 3.0 s | x7.0 |

This is a pretty big patch, and I would expect various issues to crop up during testings.
Therefore it's going to be important to test this feature thoroughly before considering building a new release around it.

Open discussions on https://github.com/lz4/lz4/issues/1334

Known current limitations : 
- Requires `pthread` (i.e. mostly `posix` environments)
- Compression only
- Independent blocks only (revert to single thread for linked blocks)
- When there are multiple files, they are still processed sequentially

Cyan4973 committed:
added a simple (untested) threadpool implementation

Cyan4973 committed:
first implementation, for compression of legacy format

Cyan4973 committed:
control over nbThreads

and automatic determination

Cyan4973 committed:
better time measurement and reportin

Cyan4973 committed:
more generic read and write jobs

no longer tied to Legacy format:
receive buffer sizes as parameters.

Cyan4973 committed:
multithreading works with lz4f format, but

- no crc
- no stdin support
- no dictionary support
- relatively wasteful resource management

Cyan4973 committed:
fix frame checksum in MT mode

Cyan4973 committed:
fixed stdin support

Cyan4973 committed:
fix legacy format compression

Cyan4973 committed:
fixed smaller block sizes

Cyan4973 committed:
preparation to support linked blocks

currently not active : behave as if blocks are still independent

Cyan4973 committed:
minor traces upgrades

- traces display line number
- errors display traces (debuglevel >= 3)

Cyan4973 committed:
lz4frame: new API: LZ4F_compressBegin_usingDict()

Note: effectively limited to using the dictionary once for now,
as opposed to once per block when blocks are independent
(no impact when blocks are linked: dictionary is supposed to be used once anyway)

Also :
- clarifies that default lz4frame block size is 64 KB
- refactor LZ4 Frame spec, dictionary paragraph
- updated manual

Cyan4973 committed:
fix minor comparator warning

Cyan4973 committed:
fix numCores detection on Linux

Cyan4973 committed:
make: fix lz4 mt compilation on linux/posix

now lz4 build defaults to single-thread mode

Cyan4973 committed:
updated Welcome message to specify if binary support multithreading or not

Cyan4973 committed:
fixed minor conversion warnings

Cyan4973 committed:
fix minor unused assignment

Cyan4973 committed:
fix minor conversion warnings

Cyan4973 committed:
fix cmake recipe

though it still builds single-threaded lz4

Cyan4973 committed:
lz4io: verbose compress/decompress operations display a summary

featuring time, speed, and cpu usage

Cyan4973 committed:
fix pedantic C90 compatibility warnings

Cyan4973 committed:
fixed several minor pedantic conversion warnings

mostly for C++98 compatibility
(which benefit is dubious, this policy could be revisited)

Cyan4973 committed:
fix another C90 pedantic warning

Cyan4973 committed:
fixed meson build

and make it build the MT version of lz4
with an option that is enabled by default but can be disabled on demand.

Also :
- added make target mesonbuild, for easy local build
- made huge tests friendlier to MT variant, for faster testing time

Cyan4973 committed:
fix minor static analysis initialization warning

Cyan4973 committed:
fix standard_variables test

Cyan4973 committed:
fix VS2010 solution

and some minor printf limitations on Windows.

Cyan4973 committed:
fix VS2022 solution

Cyan4973 committed:
fix minor static analyzer warnings

regarding uninitialized variables

Cyan4973 committed:
added a simple test for MT CLI commands

and documented the new command.

Cyan4973 committed:
make visual compilation test on appveyor less permissive

Cyan4973 committed:
fixed a few Visual Studio Static Analyzer warnings

Cyan4973 committed:
silence some clang x32 test

early versions of clang fail on a heisenbug,
on what seems to be a buggy optimizer issue (disappears with -O0).
later versions have even more issue producing basic functioning executables.
this test seems hard to setup.
might not be worth the trouble.

Cyan4973 committed:
removed cancelled declaration

fix oversight

Cyan4973 committed:
attempt to fix mingw32 compilation issues

Cyan4973 committed:
fixed MT incompatibility with --content-size

also:
provide MT-friendly time measurements for Linux/posix

Cyan4973 committed:
second attempt at fixing mingw32

Cyan4973 committed:
removed left-over trace

Cyan4973 committed:
report nb of threads in verbose mode

Cyan4973 committed:
updated manuals

Cyan4973 committed:
manually select single-thread path with -T1

Cyan4973 committed:
fix MT compatibility with dictionaries

Cyan4973 committed:
make file open errors recoverable in legacy mode

Cyan4973 committed:
fix potential leak after failure in legacy format

also:
display time/cpu statistics after legacy compression

Cyan4973 committed:
fix leak when input is an exact multiple of job size

t-mat commented on programs/util.c:
```diff
@@ -0,0 +1,243 @@
+                    }
+                } else {
+                    /* some other error */
+                    goto failed;
```
It may not happen.  But we can think about the following scenario : 
- `glpi()` returns `ERROR_INSUFFICIENT_BUFFER` at the first time.
- `buffer` is allocated, `while()` loop continues.
- `glpi()` fails at the second time. (for some reason)
- Here, potentially `buffer` will be leaked.

Cyan4973 committed:
sequential test, for easier debugging

Cyan4973 committed:
fixed initialization issue

Cyan4973 committed:
fixed potential leak scenario detected by @t-mat

and a potential double fclose()

Cyan4973 committed:
adjust tests for more parallelism

Cyan4973 committed:
minor optimization : share thread pools

In scenarios where there are many files,
instead of recreating and freeing thread pool for every file
do it once, and share it across all files.

The impact is small, but can be measured:
in a scenario compressing a lot of files of 5 MB at level 1 with 6 threads,
compression time went down from 3 sec to 2.8 sec.

Cyan4973 committed:
minor optimization: only create thread pools if needed

lz4 skips multithreading for small files.

In a scenario invoking lz4 repetitively on a lot of small files,
processing time goes down from 5.5 sec to 4.5 sec.
